### DISCUSSION TITLE
Why is django-components doing template caching on its own?

### DISCUSSION THREAD

**Discussion by oliverhaas** [django-components/django-components#879] [General]
Even with its own cache implementation etc.

Wouldn't it just be more straight forward to use the loaders given by the django settings? Per default the default template loader is cached for DEBUG=False and uncached for DEBUG=True. Also this would be probably less confusing for the user, since template caching is just... well, always handled the same.

I'm wondering if I'm missing something, or maybe this is a legacy feature or oversight in some way.

I'm not complaining about the caching; it's working.

**Comment by JuroOravec** [comment]
Which part do you find confusing? What were you trying to achieve when you came across this?

I came to Django just a year ago, and worked only on one Django project, so I don't know what may be the expectations around template caching in the wild.

My guess is that, in part, the current template caching is a legacy from what was implemented at least ~4 year ago in v0.16 [here](https://github.com/EmilStenstrom/django-components/blob/0.16/django_components/component.py#L81) (@EmilStenstrom would know better). Seems that this was then removed in [v0.26](https://github.com/EmilStenstrom/django-components/blob/0.26/django_components/component.py), but the `TEMPLATE_CACHE_SIZE` setting remained. Seeing that the `TEMPLATE_CACHE_SIZE` setting did nothing, I then brought template caching back in [v0.97](https://github.com/EmilStenstrom/django-components/releases/tag/0.97).

At that point (v0.97), django-components already supported the inlined template under `Component.template`. And this is the second reason for custom template caching - while Django's native template caching loads a template from the filesystem, for `Component.template`, we created the `Template` instance as `Template(self.template)` ([see source](https://github.com/EmilStenstrom/django-components/blob/0.96/src/django_components/component.py#L290)). And in this case the template was not being cached.

Compare that with v0.97, where we do the same, but we call our own `cached_template()` to create the instance of the `Template` ([see source](https://github.com/EmilStenstrom/django-components/blob/0.97/src/django_components/component.py#L351)). This ensured that whether you specify a template with `Component.template` or `Component.template_name`, in both cases the `Template` instances will be cached. However, for `Component.template` we used django-component's caching, while for `Component.template_name` we relied on Django's own.

---

So to sum this up:

1. django-components' template caching with `cached_template()` aims to provide the same behaviour as Django's own template caching.
2. django-components' template caching should be applied only when a template was specified with `Component.template`.

Given this, @oliverhaas I'm curious, do you use `Component.template_name` or `Component.template` to specify the template?

HOWEVER, as of next release, v0.124, the point 2. will no longer be true, and we will use `cached_template()` with both `Component.template` and `Component.template_name`. To align how the Django template is loaded with how the primary JS / CSS (`Component.js_file` / `Component.css_file`)  will be loaded.

---

>  Per default the default template loader is cached for DEBUG=False and uncached for DEBUG=True

@oliverhaas So based on the above, I think we should be able to align with Django by checking for `DEBUG`, and skip template caching when `DEBUG=True`. Could you link me to some Django documentation or source code that goes over this, so I can see how they implemented it?

**Reply by oliverhaas** [reply]
Thanks for the response, that clears it up.

The main thing I was missing was that django components also allows for directly rendering templates from strings (as you guessed I never use this), and for that it also makes sense to have some kind of cache, so the summary is basically what you wrote (square brackets my addition):

> However, for Component.template _[i.e. templates defined as strings]_ we used django-component's caching, while for Component.template_name _[i.e. templates defined as files]_ we relied on Django's own.

**Reply by oliverhaas** [reply]
Oh and one more thing:

I don't think there are any changes needed, as any setup with DEBUG=True should already reload *.py files on changes in development, so there is no need to disable or invalidate the cache on top of that.
